x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/RandomizedAssignmentRounding.java

```diff
@@ -0,0 +1,428 @@
+
+                if (m.memoryBytes() > resourceTracker.remainingNodeMemory.get(n)
+                    || m.threadsPerAllocation() > resourceTracker.remainingNodeCores.get(n)
+                    || roundedAllocations == 0
+                    || random.nextDouble() > softAssignments.get(assignment)) {
+                    unassign(assignment);
+                    assignUnderSubscribedNodes(Set.of(n));
+                } else {
+                    roundedAllocations = Math.min(roundedAllocations, resourceTracker.remainingNodeCores.get(n) / m.threadsPerAllocation());
+                    assignModelToNode(m, n, roundedAllocations);
+                    unassignOversizedModels(n);
+                    assignExcessCores(n);
+                }
+            }
+        }
+
+        private void unassignOversizedModels(Node n) {
+            for (Model m : models) {
+                Tuple<Model, Node> assignment = Tuple.tuple(m, n);
+                if (softAssignments.get(assignment) < 1.0 && m.memoryBytes() > resourceTracker.remainingNodeMemory.get(n)) {
+                    unassign(assignment);
+                }
+            }
+        }
+
+        private AssignmentPlan toPlan() {
+            AssignmentPlan.Builder builder = AssignmentPlan.builder(nodes, models);
+            for (Map.Entry<Tuple<Model, Node>, Integer> assignment : tryAssigningRemainingCores().entrySet()) {
+                builder.assignModelToNode(assignment.getKey().v1(), assignment.getKey().v2(), assignment.getValue());
+            }
+            return builder.build();
+        }
+
+        private Map<Tuple<Model, Node>, Integer> tryAssigningRemainingCores() {
+            // Eagerly assign allocations to models with larger size first on first node
+            // where the model fits.
+            //
+            // This is a trivial way to improve solution quality since increasing
+            // used allocations always improves our quality measure and we may be able to
+            // add a job, which doesn't have its quota of allocations, to the allocation
+            // random rounding finds.
+
+            Map<Tuple<Model, Node>, Integer> resultAllocations = new HashMap<>();
+
+            ResourceTracker resourceTracker = new ResourceTracker(nodes, models);
+
+            for (Model m : models) {
+                for (Node n : nodes) {
+                    Tuple<Model, Node> assignment = Tuple.tuple(m, n);
+                    int allocations = (int) Math.floor(softAllocations.getOrDefault(assignment, 0.0));
+                    resultAllocations.put(assignment, allocations);
+                    if (allocations > 0) {
+                        resourceTracker.assign(m, n, allocations);
+                    }
+                }
+            }
+
+            for (Model m : models.stream().sorted(Comparator.comparingDouble(this::remainingModelOrder)).toList()) {
+                if (resourceTracker.remainingModelAllocations.get(m) > 0) {
```

Comment:
Both of the sorted for loops (models and nodes) have a predicate check that just skips whole sections. Wouldn't it be better to filter before sorting?

Reply:
Definitely!